### PR TITLE
Noy/add api jooq

### PR DESCRIPTION
**Pull request by NoyRobinson**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -32,7 +32,7 @@ dependencies {
 	implementation("org.flywaydb:flyway-core")
 	implementation("org.flywaydb:flyway-database-postgresql")
 	implementation("org.jetbrains.kotlin:kotlin-reflect")
-	implementation("org.jooq:jooq:3.16.9")
+	implementation("org.jooq:jooq:3.17.9")
```
Comment (omermelamed) [inline build.gradle.kts]: Please separate into a different pr

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt [hunk 2]
```diff
@@ -21,11 +21,11 @@ class OwnerDao @Inject constructor(private val sql: DSLContext) {
         )
     }
 
-    fun createOwner(ownerName: String, companyId: Long, employeeId: String): Long {
+    fun createOwner(ownerInfo: Owner): Long {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt]: why did you change it?
If you have Owner meaning it was already created (have an id), you can use parameters or create a different data class for creation

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 3]
```diff
@@ -1,17 +1,13 @@
...
 import org.jooq.impl.DSL
-import org.springframework.stereotype.Component
-import java.sql.Date
-import java.util.*
-import kotlin.random.Random
-
-@Component
+import org.springframework.stereotype.Repository
 
+@Repository
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:10]: Why not @AutoWired?

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 4]
```diff
@@ -27,13 +23,13 @@ class PetDao @Inject constructor(private val sql: DSLContext) {
         )
     }
 
-    fun createPet(name: String, type: String, companyId: Long, dateOfArrival: Date, ownerId: Long?): Long {
+    fun createPet(pet: Pet): Long {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt]: Same comment as in OwnerCreation

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt [hunk 5]
```diff
@@ -48,6 +44,13 @@ class PetDao @Inject constructor(private val sql: DSLContext) {
             .and(petTable.companyId.eq(companyId)))
             .fetch(petMapper)
 
+    fun getPetById(petId: Long, companyId: Long): PetData? =
+        sql.select(petTable.id, petTable.name, petTable.type, petTable.companyId, petTable.dateOfArrival, petTable.ownerId)
+            .from(petTable)
+            .where(petTable.id.eq(petId))
+            .and(petTable.companyId.eq(companyId))
+            .fetchOne(petMapper)
+
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/PetDao.kt:53]: This function related to this pr?

File: src/main/kotlin/com/hibob/academy/dao/models.kt [hunk 6]
```diff
@@ -1,13 +1,17 @@
 package com.hibob.academy.dao
 
-import java.util.*
+import java.sql.Date
 
 data class Example(val id: Long, val companyId: Long, val data: String)
 
 data class OwnerData(val id: Long, val name: String, val companyId: Long, val employeeId: String)
 
+data class Owner(val name: String, val companyId: Long, val employeeId: String)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/models.kt]: I understand know, please rename this data class to something we can understand like OwnerCreationRequest

File: src/main/kotlin/com/hibob/academy/dao/models.kt [hunk 7]
```diff
@@ -1,13 +1,17 @@
...
+
 data class PetData(val id: Long, var name: String, val type: PetType, val companyId: Long, val dateOfArrival: Date, val ownerId: Long?)
 
+data class Pet(var name: String, val type: PetType, val companyId: Long, val dateOfArrival: Date, val ownerId: Long?)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/models.kt]: I understand know, please rename this data class to something we can understand like PetCreationRequest

Comment (omermelamed) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: Why var name and not val?

File: src/main/kotlin/com/hibob/academy/dao/models.kt [hunk 8]
```diff
@@ -1,13 +1,17 @@
...
-import java.util.*
+import java.sql.Date
 
 data class Example(val id: Long, val companyId: Long, val data: String)
 
 data class OwnerData(val id: Long, val name: String, val companyId: Long, val employeeId: String)
 
+data class Owner(val name: String, val companyId: Long, val employeeId: String)
+
 data class PetData(val id: Long, var name: String, val type: PetType, val companyId: Long, val dateOfArrival: Date, val ownerId: Long?)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/dao/models.kt]: Why var name?

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+import org.springframework.web.bind.annotation.RequestBody
+
+@Controller
+@Path("/api/noy/owners")
+@Produces(MediaType.APPLICATION_JSON)
+class OwnerResource(private val ownerService: OwnerService) {
+
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    fun addOwner(@RequestBody owner: Owner): Response {
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: do we have to use this annotation @RequestBody ?

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+        val ownerId = ownerService.createOwner(owner)
+        return Response.status(Response.Status.CREATED).entity("new owner created with id $ownerId").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: Response.ok is enough

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @GET
+    @Path("/{companyId}/getOwner")
+    fun getOwner(@PathParam("companyId") companyId: Long): Response {
+        val owner = ownerService.getOwner(companyId)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: maybe getAllOwners?

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+        return Response.status(Response.Status.OK).entity(owner).build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: Response.ok instead of Response.status(Response.Status.OK)

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    }
+
+    @GET
+    @Path("/{petId}/ownerInformation")
+    fun getOwnerForPet(@PathParam("petId") petId: Long): Response {
+        val ownerInfo = ownerService.getOwnerByPetId(petId)
+            ?: return Response.status(Response.Status.NOT_ACCEPTABLE).entity("Pet doesn't have an owner").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: Why not handle in the service and throw an exception if you didnt find any owner?
What about companyId?

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt [hunk 14]
```diff
@@ -1,78 +1,64 @@
...
 
-        // create
         @POST
         @Consumes(MediaType.APPLICATION_JSON)
-        fun addPet(@RequestBody pet: PetData): Response {
-            Response.status(Response.Status.CREATED).build()
-            return Response.ok(pet).build()
+        fun addPet(@RequestBody pet: Pet): Response {
+            val petId = petService.createPet(pet)
+            return Response.status(Response.Status.CREATED).entity("new pet created with id $petId").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetResource.kt]: Response.ok is enough

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt [hunk 15]
```diff
@@ -1,78 +1,64 @@
...
         @Consumes(MediaType.APPLICATION_JSON)
         @Path("/{petId}/updatePetName")
-        fun updateName(@PathParam("petId") petId: Long, @QueryParam("newName") newName: String?): Response {
-            val pet = PetData(id = petId, name = "Angie", type = PetType.DOG, companyId = 12L, dateOfArrival = Date(), ownerId = null)
-            pet.name = newName ?: pet.name
-            Response.status(Response.Status.ACCEPTED).build()
-            return Response.ok(pet).build()
+        fun updateName(@PathParam("petId") petId: Long, @QueryParam("newName") newName: String, @QueryParam("companyId") companyId: Long): Response {
+            petService.updatePetName(petId, newName, companyId)
+            return Response.status(Response.Status.ACCEPTED).entity("Pets name changed successfully").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetResource.kt]: Response.ok is enough

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt [hunk 16]
```diff
@@ -1,78 +1,64 @@
...
-        // http://localhost:8080/api/noy/pets/updatePetName/3983f705-e3c4-4515-b46e-b8c74936ffd9?newName=Nessy
-
+        @PUT
+        @Consumes(MediaType.APPLICATION_JSON)
+        @Path("/{petId}/{ownerId}/updatePetsOwner")
+        fun updateOwnerForPet(@PathParam("petId") petId: Long, @PathParam("ownerId") ownerId: Long, @QueryParam("companyId") companyId: Long): Response {
+            val success = petService.updatePetOwner(petId, ownerId, companyId)
+            if(success == 1)
+                return Response.status(Response.Status.OK).entity("Pet $petId got a new owner!").build()
+            return Response.status(Response.Status.BAD_REQUEST).entity("Pet doesn't exist or pet $petId already has an owner!").build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetResource.kt]: Why not handle the success in the service and throw an exception if doesnt success?

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt [hunk 17]
```diff
@@ -1,78 +1,64 @@
...
-                    pet.id == petId
-                }.isEmpty())
-                return Response.status(Response.Status.NOT_FOUND).build()
-            Response.status(Response.Status.OK).build()
-            return Response.ok("dog").build()
+        @Consumes(MediaType.APPLICATION_JSON)
+        @Path("/{petType}")
+        fun getPetType(@PathParam("petType") petType: PetType, @QueryParam("companyId") companyId: Long): Response {
+            val petsByType = petService.getPetsByType(petType, companyId)
+            return Response.status(Response.Status.OK).entity(petsByType).build()
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetResource.kt]: Nice

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt [hunk 18]
```diff
@@ -1,78 +1,64 @@
...
-class PetResource {
-
-
-    // Get pets by owner: Given an owner ID, return all the information of the pets this owner adopted.
-    // Count pets by type: This API does not receive any information but return a map of the type to how many pets of that type we have in the DB
-
-    @Controller
+@Controller
     @Path("/api/noy/pets")
     @Produces(MediaType.APPLICATION_JSON)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/resource/PetResource.kt:14]: You can put the @Consumes here as well and remove from all functions

File: src/main/kotlin/com/hibob/academy/service/OwnerService.kt [hunk 19]
```diff
@@ -0,0 +1,15 @@
...
+import jakarta.inject.Inject
+import org.springframework.stereotype.Service
+
+
+@Service
+class OwnerService @Inject constructor(private val ownerDao: OwnerDao) {
+
+    fun createOwner(owner: Owner) = ownerDao.createOwner(owner)
+    fun getOwner(companyId: Long) = ownerDao.getOwner(companyId)
+    fun getOwnerByPetId(petId: Long) = ownerDao.getOwnerByPetId(petId)
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/service/OwnerService.kt]: enter between functions

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 8 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: I did it in order to have a function that updated the pets name but Eli explained that i should change it to val and use copy in order to change the name

File: src/main/kotlin/com/hibob/academy/dao/models.kt (hunk 7 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/dao/models.kt]: same as in owner, i changed it

File: src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt (hunk 2 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/dao/OwnerDao.kt]: people in the group explained to me that it's clean code to change it

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt (hunk 11 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: i agree, ill change it

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt (hunk 9 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: Because i'm receiving a Json object of the owner

File: src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt (hunk 13 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/resource/OwnerResource.kt]: i'll fix it

File: src/main/kotlin/com/hibob/academy/resource/PetResource.kt (hunk 16 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/resource/PetResource.kt]: fixed it

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 3 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:10]: I saw that usually @Repository is used in order to indicate that a class is a DAO

File: src/main/kotlin/com/hibob/academy/dao/PetDao.kt (hunk 5 above)
Comment (NoyRobinson) [reply src/main/kotlin/com/hibob/academy/dao/PetDao.kt:53]: it wasn't in the assignment but i thought its a function that could by useful

File: src/main/kotlin/com/hibob/academy/service/OwnerService.kt [hunk 20]
```diff
@@ -0,0 +1,22 @@
...
+import org.springframework.stereotype.Service
+
+@Service
+class OwnerService @Inject constructor(private val ownerDao: OwnerDao) {
+
+    fun createOwner(owner: OwnerCreationRequest) = ownerDao.createOwner(owner)
+
+    fun getAllOwners(companyId: Long) = ownerDao.getAllOwners(companyId)
+
+    fun getOwnerByPetId(petId: Long, companyId: Long): OwnerData? {
```
Comment (NoyRobinson) [inline src/main/kotlin/com/hibob/academy/service/OwnerService.kt]: no need ?

File: src/main/kotlin/com/hibob/academy/service/PetService.kt [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+    fun getPetsByType(petType: PetType, companyId: Long) = petDao.getPetsByType(petType, companyId)
+
+    fun getPetById(petId: Long, companyId: Long): PetData? = petDao.getPetById(petId, companyId)
+
+    fun getAllPets(companyId: Long) = petDao.getAllPets(companyId)
+
+    fun updatePetOwner(petId: Long, ownerId: Long, companyId: Long) {
+        val success = petDao.updatePetOwner(petId, ownerId, companyId)
+        if(success == 1) return
+        throw IllegalStateException("Pet doesn't exist or pet already has an owner")
```
Comment (NoyRobinson) [inline src/main/kotlin/com/hibob/academy/service/PetService.kt]: > 0

File: src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt [hunk 22]
```diff
@@ -24,88 +19,70 @@ class PetDaoTest @Autowired constructor(private val sql: DSLContext){
...
-        petDao.createPet(petName1, petTypeString, companyId, dateOfArrival, 123L)
-
-        val petName2 = "Nessy"
-        petDao.createPet(petName2, petTypeString, companyId, dateOfArrival, 123L)
-
+        val pet1 = PetrCreationRequest("Angie", PetType.convertStringToPetType("Dog"), companyId, Date.valueOf("2010-05-20"), 123L)
+        petDao.createPet(pet1)
+        val pet2 = PetrCreationRequest("Nessy", pet1.type, pet1.companyId, pet1.dateOfArrival, pet1.ownerId)
+        petDao.createPet(pet2)
         val actual = petDao.getPetsByOwner(123L, companyId).size
```
Comment (NoyRobinson) [inline src/test/kotlin/com/hibob/academy/dao/PetDaoTest.kt]: fix

File: src/main/kotlin/com/hibob/academy/service/PetService.kt [hunk 23]
```diff
@@ -0,0 +1,33 @@
...
+@Service
+class PetService @Inject constructor(private val petDao: PetDao) {
+
+    fun createPet(pet: PetrCreationRequest): Long =
+        petDao.createPet(pet)
+
+    fun getPetsByType(petType: PetType, companyId: Long) =
+        petDao.getPetsByType(petType, companyId)
+
+    fun getPetById(petId: Long, companyId: Long): PetData? =
```
Comment (omermelamed) [inline src/main/kotlin/com/hibob/academy/service/PetService.kt]: why `?` why not throw an exception if null?

**Review by omermelamed — APPROVED** [review]
